index.bs

```diff
@@ -2015,14 +1902,25 @@ Issue: This needs to be generalised to work with realms too.
 <div algorithm="remote end steps for session.unsubscribe">
 The [=remote end steps=] with |session| and |command parameters| are:
 
-1. Let the |list of event names| be the value of the <code>events</code> field of
-   |command parameters|.
+1. If |command parameters| matches the session.SubscriptionRequest production:
+
+   1. Let the |list of event names| be the value of the <code>events</code> field of
+      |command parameters|.
+
+   1. Let the |list of contexts| be the value of the <code>contexts</code>
+      field of |command parameters| if it is present or an empty [=/list=] if it isn't.
+
+   1. Remove all subscriptions from |session|'s [=subscriptions=] whose event names is |list of event names|
```

Comment:
Yeah, I think this will be the tricky part with this approach, because the contract is that at the end you're no longer subscribed to any of the events for any of the contexts, but you're still subscribed to other events for those contexts, and those events for other contexts.

So we need to handle something like
```
subscribe(contexts=[1,2], events=[A, B])
subscribe(contexts=[2, 3], events=[B, C])
unsubscribe(contexts=[1, 2], events=[B])
```

Maybe it's easier if we model subscriptions as a map from a list of events to a list of contexts (plus an id ofc, but I'll ignore that for simplicity). Then you have the above stored as:
```
[
  {
    A: Contexts([1,2]),
    B: Contexts([1,2])
  },
  {
    B: Contexts([2,3]),
    C: Contexts([2,3])
  }
}
```
and it's easier to see how unsubscribe should work: you just iterate through the list of subscriptions, and for each event that's being unsubscribed you remove the contexts you're unsubscribing from the list of subscribed contexts, so you end up with:
```
[
  {
    A: Contexts([1,2]),
  },
  {
    B: Contexts([3]),
    C: Contexts([2, 3])
  }
}
```

Reply:
I would probably prefer to keep subscriptions as a struct and perhaps remove the event names and contexts from the matching subscriptions (which means mutating subscriptions so we won't be able to return subscription data as is if we ever need to) and then remove "empty" subscriptions. In general, I think this unsubscribe behavior is somewhat confusing and I wonder if we want to keep it long-term or if we want to only allow ID-based unsubscribe calls eventually?

Reply:
It feels hard to me with the struct approach because each struct has a list of events E and a list of contexts C, and all of ExC are subscribed, but once you unsubscribe you are left with only some subset that are valid, so with the you need to split into i structs with the same id so that now summing over E_i x C_i gives all the valid subscriptions.

I guess the most straightforward, but inefficient, way of doing this is to model it as a totally flat list where each entry has an id, an event, and a context, and unsubscribing means removing all the entries that either match the id or the event name / context.

I agree that this behaviour is somewhat confusing, but I think we're already in a place where it's going to be very hard to deprecate the existing behaviour completely. Maybe eventually we can do it, first by making it opt-in, but I certainly think we're going to need to design in a way that makes the current model implementable.

Reply:
At least I hope there will be a way to avoid introducing this behavior for user context subscriptions? 

> hard to deprecate the existing behaviour completely.

Probably, although Puppeteer, for instance, does not use unsubscribe at all at the moment so perhaps no many clients had adopted edge cases for unsubscribe. A simpler way would be to deprecate the ability to modify subscriptions via unsubscribe: could we instead make a change that unsubscribing only removes the subscriptions that exactly match the subscribe requests? 

I would probably attempt the split upon unsubscribing first with allowing duplicate subscription IDs in the list.

Reply:
> could we instead make a change that unsubscribing only removes the subscriptions that exactly match the subscribe requests? 

> subscribe(contexts=[1,2], events=[A, B])
> subscribe(contexts=[2, 3], events=[B, C])
> unsubscribe(contexts=[1, 2], events=[B])

with such change the unsubscribe call here would do nothing (we could throw an error if nothing matched). This would allow probably most of the clients to work in almost the same way as I expect most unsubscribe requests to be the same as subscribe request.

Reply:
Requiring subscriptions to exactly match the subscribe requests feels like very breaking behaviour. For example if you have tests with a `subscribe()` command that adds the events / contexts to sets and a teardown fixture that does a single unsubscribe from all events at the end of the test.

Incidentally, for the same reason, we should allow unsubscribing from multiple ids at the same time.

I really think we should try as hard as we can to keep the existing model working for now. I don't even think it's too hard. I'd be totally happy to start emitting warnings for unsubscribe requests that don't use ids once the new model is implemented.

Reply:
Interestingly, even our own WPT fixtures seem to unsubscribe in a symmetrical way https://github.com/web-platform-tests/wpt/blob/22b20cf0eb577a7df17f7105e47e2b1b818d07b3/webdriver/tests/support/fixtures_bidi.py#L64

I think we could still treat empty contexts set as matching "any" subscription to allow global unsubscribe (and aligned with the fields in the subscription struct). I mean I will try to keep the existing behavior but I wonder if anyone is actually using it at the moment (otherwise, it might be just extra complexity for the spec).

Selenium seems to be only doing global unsubscribes https://github.com/search?q=repo%3ASeleniumHQ%2Fselenium+session.unsubscribe+language%3AJava&type=code&l=Java

Reply:
for the case of 

```
subscribe(events=[A])
unsubscribe(contexts=[1, 2], events=[A])
```

the current behavior seems to be to do nothing, right?